### PR TITLE
8330 Correct issue where permission umask of files were not set when …

### PR DESCRIPTION
**Pull request by jimmy-ho**
…archiving in zip format with ZipArchiver

### CODE REVIEW

File: src/Composer/Package/Archiver/ZipArchiver.php [hunk 1]
```diff
@@ -44,6 +44,18 @@ public function archive($sources, $target, $format, array $excludes = array(), $
                     $zip->addEmptyDir($localname);
                 } else {
                     $zip->addFile($filepath, $localname);
+
+                    /**
+                     * ZipArchive::setExternalAttributesName is available from >= PHP 5.6
+                     */
+                    if (defined('PHP_VERSION_ID') && PHP_VERSION_ID >= 50600) {
```
Comment (stof) [inline src/Composer/Package/Archiver/ZipArchiver.php]: `defined('PHP_VERSION_ID')` will always be true (it was added in some PHP 5.2 version)

Comment (jimmy-ho) [reply src/Composer/Package/Archiver/ZipArchiver.php]: Right, I saw similar checks so added it for safety

**Comment by jimmy-ho** [comment]
> should we also set the permissions when adding an empty dir ?

Guess there should not be any harm to

**Comment by jimmy-ho** [comment]
@stof code updated to always be applied on directory and files, and removed the defined() check if the minimum requirement is PHP 5.3 for composer.

**Comment by stof** [comment]
Have you check the support of empty dirs here ? According to https://stackoverflow.com/questions/36287554/php-ziparchive-file-permissions you might need to handle dirs with a specific change.

**Comment by jimmy-ho** [comment]
> Have you check the support of empty dirs here ? According to https://stackoverflow.com/questions/36287554/php-ziparchive-file-permissions you might need to handle dirs with a specific change.

@stof I have not had this use-case to test unfortunately, the immediate issue has been the executable flag for files being dropped... this has caused binaries which should be executable to stop working.

Happy to explore if there are examples to really validate that use-case - perhaps a new issue to address it further.

**Comment by stof** [comment]
Well, try to run the archiver on a folder containing an empty dir and see whether the empty dir preserves the permission when unzipping.
I don't think we have existing issues about permissions of empty directories specifically, given that the existing archiver does not support permissions at all. But adding support for permissions should ideally be done in a working way

**Comment by alcohol** [comment]
Do we need to take windows hosts into account here, or can we safely run this at all times?

**Comment by a-camacho** [comment]
Any news from this ? We are also experiencing the problem where composer is installing packages and using CHMOD 666 instead of default value of server CHMOD 644. When I unzip the .zip or .tar file myself it works, but not with Composer.

**Comment by alcohol** [comment]
@a-camacho this issue is unrelated to yours. You are referring to Composer extracting files from archive(s). This issue is related to archives generated by Composer.

**Comment by stof** [comment]
@alcohol he's referring to extracting files from a Satis repo according to https://github.com/composer/satis/issues/577, and so extracting files generated by the code being fixed here.

**Comment by Seldaek** [comment]
Thanks

**Pull request merged**